### ISSUE TITLE
Crashes due to updating or dismounting components with nil roots

### ISSUE THREAD

**Issue by ZackaryWelch** [maxence-charriere/go-app#652]
I've noticed a series of crashes that I thought I had pinned down, but now I'm completely stuck.

I've added logging to two different places in component.go: updateRoot() and dismount().

Errors occur when the root, a, is nil, whereupon update(a, b) crashes. Dismount has the same issue, with c.root being nil, causing dismount(c.root) or c.ctxCancel() to crash. Besides the fact this logging should be present, as otherwise it's impossible to pinpoint what's causing the crash, the reason these crashes occur is still unknown to me. I thought initially it was due to pointers being allocated in the wrong place, but creating a series of components in a map in OnInit and then switching between those components causes a crash.

Here's the initalization in OnInit:
```
    C.licenseDetails = make(map[string]*CompLicenseDetails)
    for _, id := range licenseIDs {
      C.licenseDetails[id] = &CompLicenseDetails{
        LicenseID: id,
        Popup:     C.Popup,
      }
    }
```
CompLicenseDetails being its own component with its own OnInit function. Too switch between them, I have a series of links in the parent component that switch the body of a static popup component. To switch, I replace an attribute of the popup component and then call update on the component. Calling a direct update is needed or else nothing will happen. Components being exported or non-exported fields doesn't appear to have any impact.

In summary, my issue is:

1) Why is this happening?

2) Is this intended?

3) Can there be better logging to show which component is causing the crash and why?

4) Is there another way to switch out a child component and update?

**Comment by oderwat** [comment]
I did not really read into what you have written. But it may have to do with a problem I ran into in the past: Components of the same type don't get switched (dismounted/remounted) but only "updated". I created [this](https://github.com/metatexx/go-app-pkgs) to work around this.

**Comment by ZackaryWelch** [comment]
@oderwat Thanks for the reply and the work you put into that project. I made use of it to fix this problem, although I ran into an additional issue using it much like this:
`Failed to execute 'replaceChild' on 'Node': The node to be replaced is not a child of this node.`

However, I fixed it by changing replaceChild in js_wasm to replaceWith, and now it works perfectly. replaceWith is a bit newer but as long as you don't need IE support it's fine. I might either make a PR or keep it as a fork.
